In [1]:
import mpp_classifiers as mpp
import load_ship_data as lsd
import numpy as np
import time
from sklearn.decomposition import PCA

In [2]:
data_path="C:/sassassas_data/archive/shipsnet.json"
train,test,valid=lsd.load_data_train_test_split(data_path)

Training Set Data Length:  2800   Label Length:  2800
TestingSet Set Data Length:  600  Label Length:  600
Validation Set Data Length:  600  Label Length:  600


In [3]:
ship_prob=np.sum(train[1])*1.0/train[1].shape[0]
p=np.array([1-ship_prob,ship_prob])
print(p)
Xtrain=train[0]
ytrain=train[1]
Xtest=test[0]
ytest=test[1]
Xtrain=Xtrain.reshape(Xtrain.shape[0],Xtrain.shape[1]*Xtrain.shape[2]*Xtrain.shape[3])
Xtest=Xtest.reshape(Xtest.shape[0],Xtest.shape[1]*Xtest.shape[2]*Xtest.shape[3])
print(Xtrain.shape)
print(Xtest.shape)

[0.74535714 0.25464286]
(2800, 19200)
(600, 19200)


In [4]:
t0 = time.time()           # start time
y_model = mpp.mpp(Xtrain, ytrain, Xtest, 1, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')

Overall accuracy = 0.7566666666666667;
Classwise accuracy = [1. 0.];
The learning process takes 146.8991458415985 seconds.


In [5]:
mpp_1_0_p=p
mpp_1_0_acc=acc_overall
pmm_1_0_class=acc_classwise
mpp_1_0_time=t1-t0
p=np.array([.5,.5])
t0 = time.time()           # start time
y_model = mpp.mpp(Xtrain, ytrain, Xtest, 1, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')
mpp_1_1_p=p
mpp_1_1_acc=acc_overall
pmm_1_1_class=acc_classwise
mpp_1_1_time=t1-t0

Overall accuracy = 0.685;
Classwise accuracy = [0.6938326  0.65753425];
The learning process takes 42.332788705825806 seconds.


In [6]:
p=np.array([.9,.1])
t0 = time.time()           # start time
y_model = mpp.mpp(Xtrain, ytrain, Xtest, 1, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')
mpp_1_2_p=p
mpp_1_2_acc=acc_overall
pmm_1_2_class=acc_classwise
mpp_1_2_time=t1-t0

Overall accuracy = 0.7566666666666667;
Classwise accuracy = [1. 0.];
The learning process takes 41.83013391494751 seconds.


In [4]:
pca = PCA(n_components=1920)
pca.fit(Xtrain)

PCA(n_components=1920)

In [6]:
pca_Xtrain=pca.transform(Xtrain)

In [7]:
pca_Xtest=pca.transform(Xtest)

In [16]:
explained=np.sum(pca.explained_variance_ratio_)
print(explained)

0.9989776178004813


In [8]:
p=np.array([1-ship_prob,ship_prob])
t0 = time.time()           # start time
y_model = mpp.mpp(pca_Xtrain, ytrain, pca_Xtest, 1, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')

Overall accuracy = 0.7583333333333333;
Classwise accuracy = [1. 0.];
The learning process takes 0.32712292671203613 seconds.


In [10]:
pca2 = PCA(n_components=192)
pca2.fit(Xtrain)

PCA(n_components=192)

In [15]:
pca2_Xtrain=pca2.transform(Xtrain)
pca2_Xtest=pca2.transform(Xtest)
explained=np.sum(pca2.explained_variance_ratio_)
print(explained)

0.9391609273657975


In [18]:
p=np.array([1-ship_prob,ship_prob])
t0 = time.time()           # start time
y_model = mpp.mpp(pca2_Xtrain, ytrain, pca2_Xtest, 1, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')

Overall accuracy = 0.7583333333333333;
Classwise accuracy = [1. 0.];
The learning process takes 0.028922557830810547 seconds.


In [19]:
p=np.array([1-ship_prob,ship_prob])
t0 = time.time()           # start time
y_model = mpp.mpp(pca2_Xtrain, ytrain, pca2_Xtest, 2, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')

Overall accuracy = 0.93;
Classwise accuracy = [0.96263736 0.82758621];
The learning process takes 0.8517212867736816 seconds.


In [20]:
p=np.array([1-ship_prob,ship_prob])
t0 = time.time()           # start time
y_model = mpp.mpp(pca2_Xtrain, ytrain, pca2_Xtest, 3, p)
t1 = time.time()           # ending time
acc_classwise, acc_overall = mpp.accuracy_score(ytest, y_model)
print(f'Overall accuracy = {acc_overall};')
print(f'Classwise accuracy = {acc_classwise};')
print(f'The learning process takes {t1 - t0} seconds.')

C:\Users\Syste\anaconda3\lib\site-packages\numpy\linalg\linalg.py:2159: RuntimeWarning: overflow encountered in det
  r = _umath_linalg.det(a, signature=signature)


Overall accuracy = 0.7583333333333333;
Classwise accuracy = [1. 0.];
The learning process takes 1.1858265399932861 seconds.


Observations for Report:

Even with a linear classifier, our flattened training data set of dimension [2800,19200] takes an excessive amount of time to classify. Euclidean distance (case 1) yielded an accuracy of 75.6667%. Classwise accuracy showed that case 1 was merely classifying all samples as class 0 (no ship). The training time on this, our fastest classifier, was 146.8991 seconds. Ergo, PCA and FLD will be critical to classifying this data in a timely fashion. Using PCA with 1920 dimensions (99.898% of variance explained) reduced the training time for case 1 from 146.8991 seconds to .3271 seconds without any loss of accuracy. However, this dimension reduction was insufficient for case 2, which still took an excessive amount of time to train. So we ran PCA again with dimension 192. This preserved 93.916% of the data, and yielded extremely quick classification by all three cases. Mahalanobis distance (case 2) performed the best by a large margin with 93% accuracy. Classwise performace using Mahalanobis distance was also vastly improved from Euclidean distance with 96.2637% accuracy on class 0 (no ship) and 82.7586% accuracy on class 1 (ship). Quadratic distance performed the same as Euclidean distance (75.8333% accuracy), classifying all inputs as class 0 (no ship). 

Major Results: PCA+Mahalanobis yielded 93% accuracy with excellent classwise accuracy (96%,83%) and runtime (.85 seconds).
               Dimension reduction is critical on dataset of this size.